eth/downloader/queue.go

```diff
@@ -210,96 +310,160 @@ func (q *queue) Insert(headers []*types.Header, from uint64) []*types.Header {
 			glog.V(logger.Warn).Infof("Header #%v [%x] broke chain ancestry", header.Number, hash[:4])
 			break
 		}
-		// Queue the header for body retrieval
+		// Make sure no duplicate requests are executed
+		if _, ok := q.blockTaskPool[hash]; ok {
+			glog.V(logger.Warn).Infof("Header #%d [%x] already scheduled for block fetch", header.Number.Uint64(), hash[:4])
+			continue
+		}
+		if _, ok := q.receiptTaskPool[hash]; ok {
+			glog.V(logger.Warn).Infof("Header #%d [%x] already scheduled for receipt fetch", header.Number.Uint64(), hash[:4])
+			continue
+		}
+		// Queue the header for content retrieval
+		q.blockTaskPool[hash] = header
+		q.blockTaskQueue.Push(header, -float32(header.Number.Uint64()))
```

Comment:
It's required by [my priority queue implementation](https://godoc.org/gopkg.in/karalabe/cookiejar.v2/collections/prque). The idea was that computationally and memory wise it costs the same as an int, but gives a greater flexibility if you need to insert items into the queue after already having a bunch in there. Not sure it it's a good or bad decision, but it's here to stay for now :P